align_system/algorithms/outlines_regression_adm.py

```diff
@@ -0,0 +1,312 @@
+        choices = [a.unstructured for a in available_actions]
+
+        if len(set(choices)) != len(choices):
+            log.warning("Unstructured text for available actions is not "
+                        "unique, appending action parameters to choices")
+
+            character_id_to_name = {c.id: c.name for c in scenario_state.characters}
+            # Important that the choices stay in the same order as the
+            # available actions as we'll use the selected index later to
+            # map to the corresponding action
+            choices = []
+            for a in available_actions:
+                if(a.action_type == ActionTypeEnum.APPLY_TREATMENT
+                   and a.parameters is not None and len(a.parameters) > 0):
+                    choices.append(detailed_unstructured_treatment_action_text(a, character_id_to_name))
+                elif(a.action_type == ActionTypeEnum.TAG_CHARACTER
+                     and a.parameters is not None and len(a.parameters) > 0):
+                    choices.append(detailed_unstructured_tagging_action_text(a, character_id_to_name))
+                else:
+                    # Not covering every possible case here, may need
+                    # to add more dedicated detailed prompts
+                    choices.append(a.unstructured)
+
+        target_kdmas = alignment_target.kdma_values
+
+        # Get kdma names and descriptions 
+        with open(KDMA_DESCRIPTIONS_FILE_PATH, 'r') as f:
+            kdma_descriptions = yaml.load(f, Loader=yaml.FullLoader)
+        # Add names and descriptions to target_kdmas
+        for kdma_idx in range(len(target_kdmas)):
+            if not isinstance(target_kdmas[kdma_idx], dict):
+                if isinstance(target_kdmas[kdma_idx], kdma_value.KDMAValue):
+                    target_kdmas[kdma_idx] = target_kdmas[kdma_idx].to_dict()
+                else:
+                    target_kdmas[kdma_idx] = dict(target_kdmas[kdma_idx])
+            kdma = target_kdmas[kdma_idx]['kdma']
+            if kdma not in kdma_descriptions:
+                raise RuntimeError("Missing target kdma description.")
+            else:
+                target_kdmas[kdma_idx]['name'] = kdma_descriptions[kdma]['name']
+                target_kdmas[kdma_idx]['description'] = kdma_descriptions[kdma]['description']
+
+        predicted_kdma_values = {}      # Dictionary to save output for each choice to
+        for choice in choices:
+            predicted_kdma_values[choice] = {
+                'outcomes':[]           # List to add sampled outcome predictions to
+            }
+            for target_kdma in target_kdmas:
+                predicted_kdma_values[choice][target_kdma['kdma']] = {
+                    'scores':[],        # List to add sampled kdma scores to
+                    'reasonings':[]     # List to add sampled kdma score reasonings to
+                }
+
+        # Sample multiple predictions
+        for _ in range(num_samples):
```

Comment:
Outlines can handle this for us by passing in multiple prompts at a time in a list, unless there's some other compelling reason to use a loop here (we don't have to refactor this now for this MR, but a potential optimization)

Reply:
Refactored in [3a6f9f6](https://github.com/ITM-Kitware/align-system/pull/62/commits/3a6f9f61b1a557598f8e2cf0d58b8597a3e9a765)